### PR TITLE
Added Monte Carlo simulation for winning odds

### PR DESCRIPTION
**Pull request by asegs**
This introduces a second public method on `poker-evaluator`for estimating win probability using a Monte Carlo simulation.  It is explicitly modeled for Texas Hold 'Em.

The API takes the hole cards the player has, the current community cards (if any), the number of players at the table, and the number of trials to run.  The result is a 0-1 probability that the player will win the hands.  

This is a fast and simple implementation of this mechanic.  Performance is influenced mainly by the number of cycles multiplied by the number of players, but also impacted by the number of community cards dealt.  

While a higher number of cycles doesn't lead to dramatically different evaluation, this implementation is capable of considering about 4 million player hands (player count * cycles) per second on my machine.

Example usage:
```javascript
winningOdds(['ah','as'],[],5, 1000) // 0.581

winningOdds(['ah','as'],['2c','3c','4c','5c'],5, 1000) // 0.03

winningOdds(['ah','as'],[],2, 1000) // 0.846
```

### CODE REVIEW

**Comment by asegs** [comment]
@Sukhmai let me know if you want to talk about any of these changes.  I only started working on this after writing some poker evaluation tools of my own and wanted to contribute to this project.

**Comment by asegs** [comment]
And I think the added royal flush rank might not be a good idea since users of this package might have explicit handling for each type of hand.

File: src/poker-evaluator.ts
Comment (Sukhmai) [inline src/poker-evaluator.ts:29]: Small question: why did you invert this if else?

File: src/poker-evaluator.ts
Comment (Sukhmai) [inline src/poker-evaluator.ts]: Thanks for calling this out in [your comment](https://github.com/Sukhmai/poker-evaluator/pull/2#issuecomment-1828755662). I definitely agree let's revert this change because we don't want to create breaking changes for our users. This also broke two of the unit tests that expected a "straight flush" hand. Run `npm test` to see. You can handle the royal flush case in your application by looking for `handType == 9` and `handRank == 10`.

File: src/poker-evaluator.ts
Comment (Sukhmai) [inline src/poker-evaluator.ts:188]: Can you add a comment mentioning this is using https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm, just for people's reference.

File: src/poker-evaluator.ts
Comment (Sukhmai) [inline src/poker-evaluator.ts]: Stylistic nit: please remove extra spaces in all for loops. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for for stylistic conventions.

File: src/poker-evaluator.ts
Comment (Sukhmai) [inline src/poker-evaluator.ts]: Remove extra spaces also for iterators.

**Review by Sukhmai — CHANGES_REQUESTED** [review]
First of all thanks for pinging me, I didn't get an email when the PR was created probably because I wasn't watching the repository. Apologies for the slightly delayed response.

Secondly thanks for adding this to the library! I remember using http://www.propokertools.com/ a lot back in the day for calculating hand odds (I guess it's down now?), so a very fun addition. It would be cool if you could optionally specify an array of hands for opposing players as well, but obviously not necessary for this PR.

The changes requested are pretty minor, the only other thing I have to ask is if you could add some unit tests for the `winningOdds()` method to the `poker-evaluator.spec.ts` file. At a cursory glance it seems like you can't seed `Math.Random() in Javascript, so you could assert a range of values, or find a way to set the randomness component.

File: src/poker-evaluator.ts
Comment (asegs) [reply src/poker-evaluator.ts:29]: Small performance reasons, the Monte Carlo analysis in `winningOdds()` uses numbers since they are a bit faster than text.  Since the odds estimator hits this method so many times, I flipped these to make the first case we test for the one that would be true when using the odds calculator.

This is only relevant since there is a small but not completely insignificant performance overhead to testing the type of each card element not once but twice (`not string...not string...not string...yes number...yes number...yes number`) in the every function on the cards.

Comment (Sukhmai) [reply src/poker-evaluator.ts:29]: It's fine to keep it inverted then! I was just curious and I think the performance gain for the winningOdds() function is worth it, as regular users are very unlikely to call evalHand() w/ strings at the same frequency.

**Comment by asegs** [comment]
I wrote a lot here previously about variance of probability (which you can read with the previous versions) but I am retracting it and the variance related commits since after a lot of testing, it seems that variance is a direct function of probability and the extra calculation is not necessary.

Hands with winning odds of around 0.5 have the highest variance in accuracy of assessments, while hands close to 0.1 or 0.9 have the lowest variance in assessment accuracy.  This seems pretty obvious when said that way.  It is very useful for testing but not worth including as a core part of the evaluator.  I will be retracting those commits but the work has been interesting.

**Review by Sukhmai — APPROVED** [review]
Approved! Thanks for all the work, and thanks for being patient, it's been a busy week. I'll version this and update the npm package once merged. If you're building something using this tool let me know when it's finished, I'm curious!

**Comment by asegs** [comment]
Hey @Sukhmai, no worries, I'm busy as well.  Since this has no rush, I'm adding tests now, and I would like to add two features, providing an alternate version that accepts some known cards in other players hands, and giving information on tie odds as well, ie. returning an object like:
```json
{
"win": 0.4,
"tie-2": 0.15,
"tie-3": 0.1,
"tie-n":0.0
}
```
where `n` is the player count.  I'll let you know before I merge that so you can take a look.

**Comment by asegs** [comment]
Alright, this is in a state I'm happy with.  I just want to update the docs to reflect the new signatures, but I've got more information per player on split pots of various split sizes as well as a second function that takes known table information, such as players voluntarily revealing a card, all-ins, and the spectator view where all cards are known.  

Happy holidays and I hope you're getting some good time off from work!

**Comment by asegs** [comment]
@Sukhmai added in types, updated docs. Merge it when you're ready!  I'll let you know what if anything I end up doing with this.

**Comment by Sukhmai** [comment]
Merged, thanks for adding the ability to input other player's cards and distinguishing the different outcomes. This is really good work. I'll update the npm package shortly. Happy Holidays!

**Comment by asegs** [comment]
Hey @Sukhmai, any plans to update the npm package?  I haven't seen the newest version on there.  Hope you've been well!

**Comment by Sukhmai** [comment]
Hi @asegs, extreme apologies for delaying responding to this for so long. I don't really have an excuse.

I'm unable to publish the release, because when I run `npm run test`. I get a bunch of test failures. I tried debugging it for like an ~ 1 hour a couple months ago, but it seems like it's an issue of what's being passed into the players array for the tests is null. I'll try to resolve this issue and release the package shortly. 

Here's the gist if you're curious: https://gist.github.com/Sukhmai/7cae85dedf8a089e8671a6ccb7906874

**Comment by asegs** [comment]
Hey @Sukhmai, sorry about that, I realize I made some last minute type changes that I never tested.  PR: https://github.com/Sukhmai/poker-evaluator/pull/3

**Pull request merged**
